### PR TITLE
Add core logger package

### PR DESCRIPTION
**Pull request by jordangarcia**
## Summary
- Add `js-sdk-logging` package that supplies a consistent logging interface with error handling
- Add `js-sdk-utils` which are only utils, this package is meant to eventually replace the `utils/fns.js` in the `optimizely-sdk` package

This allows us to have other packages, such as the DatafileManager that leverage the LogFactory singleton.

[Design doc for Core Logger Improvements](https://docs.google.com/document/d/1rQ13WS0quy9NreX1xLQVpSHYFiY_oyqcNDISgALVyY8/edit#heading=h.6jynaot9cbnq)

## Test plan
- Unit tests

### CODE REVIEW

File: packages/utils/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
+import * as uuid from 'uuid'
```
Comment (mjc1283) [inline packages/utils/src/utils.ts]: Several exports of the `utils` module, such as `groupBy` and `objectValues`, are unused. Why are we adding them?

File: packages/utils/jest.config.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  // "roots": [
+  //   "./src"
+  // ],
+  "transform": {
+    "^.+\\.tsx?$": "ts-jest"
```
Comment (mjc1283) [inline packages/utils/jest.config.js]: Why `tsx`?

File: packages/utils/src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+export function getTimestamp(): number {
+  return new Date().getTime()
+}
+
+export function generateUUID(): string {
+  return uuid.v4()
+}
+
+/**
+ * Valids a valid is a valid TypeScript enum
```
Comment (mjc1283) [inline packages/utils/src/utils.ts]: Nit: grammar

File: packages/utils/__tests__/utils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  })
+
+  describe('objectValues', () => {
+    it('should return object values', () => {
+      expect(objectValues({ foo: 'bar', bar: 123 })).toEqual(['bar', 123])
+    })
+    // TODO test for enumerable properties only
+  })
+
+  describe('objectValues', () => {
```
Comment (mjc1283) [inline packages/utils/__tests__/utils.spec.ts]: Two `describe` blocks here for `'objectValues'` - mistake?

File: packages/logging/test.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+
```
Comment (mjc1283) [inline packages/logging/test.js]: These seem more like code samples or documentation than a test - should these live elsewhere?

File: packages/logging/src/logger.ts [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+  ERROR: 4,
+}
+
+function coerceLogLevel(level: any): LogLevel {
+  if (typeof level !== 'string') {
+    return level
+  }
+
+  level = level.toUpperCase()
+  if (level === 'WARN') {
```
Comment (mjc1283) [inline packages/logging/src/logger.ts]: I see how this coercion is nice for usability, but it seems weird that we are treating `'WARN'` as a special case. Why `'WARN`' and not also `'ERR'`?

File: packages/logging/src/models.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  info(message: string | Error, ...splat: any[]): void
+
+  debug(message: string | Error, ...splat: any[]): void
+
+  warn(message: string | Error, ...splat: any[]): void
+
+  error(message: string | Error, ...splat: any[]): void
+}
+
+export interface LogManager {
```
Comment (mjc1283) [inline packages/logging/src/models.ts]: If there is only one class implementing `LogManager`, and there only will ever be one, what do we gain by declaring this interface?

File: packages/logging/src/logger.ts [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+  }
+
+  if (!stringToLogLevel[level]) {
+    return level
+  }
+
+  return stringToLogLevel[level]
+}
+
+type LogData = {
```
Comment (mjc1283) [inline packages/logging/src/logger.ts]: Why is `LogData` declared as a type alias, as opposed to an interface? Same question also for other uses of `type` in this module.

File: packages/logging/src/logger.ts [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+  private internalLog(level: LogLevel, data: LogData): void {
+    if (!globalLogHandler) {
+      return
+    }
+
+    if (level < globalLogLevel) {
+      return
+    }
+
+    // TODO turn this into a format log function
```
Comment (mjc1283) [inline packages/logging/src/logger.ts]: What does it mean to turn this into a format log function?

File: packages/logging/__tests__/logger.spec.ts [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+      logger = getLogger()
+    })
+
+    describe('setLogLevel', () => {
+      it('should coerce "debug"', () => {
+        setLogLevel('debug')
+        expect(getLogLevel()).toBe(LogLevel.DEBUG)
+      })
+
+      it('should coerce "deBug"', () => {
```
Comment (mjc1283) [inline packages/logging/__tests__/logger.spec.ts]: Nit: test doesn't match description

File: packages/logging/src/logger.ts (hunk 6 above)
Comment (jordangarcia) [reply packages/logging/src/logger.ts]: `warn` is the standard.  for instance `console.warn`.

File: packages/logging/src/logger.ts (hunk 8 above)
Comment (jordangarcia) [reply packages/logging/src/logger.ts]: It's not going to be extended or reused outside of this file.  My general rule of thumb is to use interfaces if they are going to be extended.

File: packages/logging/src/logger.ts (hunk 9 above)
Comment (jordangarcia) [reply packages/logging/src/logger.ts]: it's already done, will remove the TODO

File: packages/logging/src/models.ts (hunk 7 above)
Comment (jordangarcia) [reply packages/logging/src/models.ts]: I think the idea of making  `LogManager` something replaceable is something we might want to do in the future.

Also the three main components are `LogManager` `LogHandler` and `LoggerFacade` so it seemed like good documentation to have them here.

File: packages/logging/test.js (hunk 5 above)
Comment (jordangarcia) [reply packages/logging/test.js]: illremove

File: packages/utils/__tests__/utils.spec.ts (hunk 4 above)
Comment (jordangarcia) [reply packages/utils/__tests__/utils.spec.ts]: will remove

File: packages/utils/jest.config.js (hunk 2 above)
Comment (jordangarcia) [reply packages/utils/jest.config.js]: this is jsut copy/pasted

Doesn't hurt to have thsi in here though

File: packages/utils/src/utils.ts (hunk 1 above)
Comment (jordangarcia) [reply packages/utils/src/utils.ts]: They are used in other packages that we will use in the near future.

File: packages/utils/src/utils.ts (hunk 3 above)
Comment (jordangarcia) [reply packages/utils/src/utils.ts]: hahahahaha

File: packages/logging/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
+{
+ "extends": "../../tsconfig.json",
+ "compilerOptions": {
+   "outDir": "./lib",
+   "noImplicitAny": true,
```
Comment (mjc1283) [inline packages/logging/tsconfig.json]: Don't need `noImplicitAny` and `strictNullChecks` here - same as the parent config.

File: tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    // "removeComments": true,                /* Do not emit comments to output. */
+    // "noEmit": true,                        /* Do not emit outputs. */
+    "suppressImplicitAnyIndexErrors": true,
+    "importHelpers": true /* Import emit helpers from 'tslib'. */,
+    // "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
+    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
+
+    /* Strict Type-Checking Options */
+    "strict": true /* Enable all strict type-checking options. */,
+    "noImplicitAny": true /* Raise error on expressions and declarations with an implied 'any' type. */,
```
Comment (mjc1283) [inline tsconfig.json]: I think if `"strict"` is `true` it has the same effect of setting all of `noImplicityAny`, `strictNullChecks`, etc. (everything listed here under Strict Type-Checking Options) to true.

**Pull request merged**